### PR TITLE
ref(metrics_indexer): Use rapidjson, dont use deepcopy

### PR DESCRIPTION
**Pull request by untitaker**
rapidjson apparently wasn't used to serialize the message because it
couldn't deal with converting non-str keys of a dict to strings. In
order to use rapidjson, produce strings in the first place.

deepcopy is probably slow (no idea by how much, but it does invoke parts
of pickle under the hood), and it seems that in the particular way we
mutate `new_payload_value`, we don't actually need it. A `dict` clone
can be very fast in comparison. It seems that we could also get away
with not copying the outer dictionary either but that's probably too
risky.

A completely made up benchmark is attached, you need to install
pytest-benchmark for it to run. Output from master:

```
---------------------------------------------------------- benchmark: 1 tests ----------------------------------------------------------
Name (time in us)                    Min         Max      Mean    StdDev    Median       IQR  Outliers  OPS (Kops/s)  Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------
test_process_messages_bench     742.5840  2,107.9940  855.4124  126.3961  824.1645  107.4000     79;52        1.1690     718           1
----------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
==================================== 1 passed, 11 deselected in 1.77s ====================================
```

...and output from this PR:

```
---------------------------------------------------------- benchmark: 1 tests ----------------------------------------------------------
Name (time in us)                    Min         Max      Mean    StdDev    Median       IQR  Outliers  OPS (Kops/s)  Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------
test_process_messages_bench     684.7050  1,582.7150  806.7605  126.6003  765.8510  123.3245     38;14        1.2395     325           1
----------------------------------------------------------------------------------------------------------------------------------------
```

### CODE REVIEW

File: tests/sentry/sentry_metrics/test_multiprocess_steps.py [hunk 1]
```diff
@@ -286,6 +286,16 @@ def test_process_messages(mock_indexer) -> None:
...
 
 
+try:
+    import pytest_benchmark  # type: ignore  # noqa
+except ImportError:
+    pass
+else:
+
+    def test_process_messages_bench(benchmark):
+        benchmark(test_process_messages)
```
Comment (fpacifici) [inline tests/sentry/sentry_metrics/test_multiprocess_steps.py:296]: Is this intentional (merging the benchmark code) or is it just left over code ?

File: src/sentry/sentry_metrics/multiprocess.py [hunk 2]
```diff
@@ -457,18 +456,18 @@ def process_messages(
                 logger.info("process_message.offset_skipped", extra={"offset": message.offset})
                 continue
             parsed_payload_value = parsed_payloads_by_offset[message.offset]
-            new_payload_value = deepcopy(parsed_payload_value)
+            new_payload_value = dict(parsed_payload_value)
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply src/sentry/sentry_metrics/multiprocess.py:459]: @untitaker  Looking at the unit test, it seems it is doing a shallow copy. 
https://github.com/getsentry/sentry/blob/master/tests/sentry/sentry_metrics/test_multiprocess_steps.py#L242

Could you please try to replace the shallow with a deep copy in the unit test and ensure it still passes.

[the remark this answers is not in the document]
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/multiprocess.py:459]: I honestly thought I did run into issues without deepcopy, but maybe that was back when doing multiprocessing

File: tests/sentry/sentry_metrics/test_multiprocess_steps.py (hunk 1 above)
Comment (untitaker) [reply tests/sentry/sentry_metrics/test_multiprocess_steps.py:297]: it's intentional but I guess kind of silly to keep dead code in the testsuite. I can remove it if you'd like

File: src/sentry/sentry_metrics/multiprocess.py [hunk 3]
```diff
@@ -456,19 +455,18 @@ def process_messages(
             if message.offset in skipped_offsets:
                 logger.info("process_message.offset_skipped", extra={"offset": message.offset})
                 continue
-            parsed_payload_value = parsed_payloads_by_offset[message.offset]
-            new_payload_value = deepcopy(parsed_payload_value)
+            new_payload_value = parsed_payloads_by_offset.pop(message.offset)
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply src/sentry/sentry_metrics/multiprocess.py:458]: Let's ensure Meredith's PR gets merged first. 
If we overwrite a message we loose one message. Bad but manageable.
If we see a KeyError the consumer crashes and has chances to get stuck if the same partitions end up on the same consumer.

**Comment by untitaker** [comment]
It appears this gives us 2k msgs/s on top in the testenv:

Screenshot 2022-06-08 at 18 25 18

Screenshot 2022-06-08 at 18 25 26

compared to baseline:

Screenshot 2022-06-08 at 18 26 03

Screenshot 2022-06-08 at 18 26 09

File: tests/sentry/sentry_metrics/test_multiprocess_steps.py (hunk 1 above)
Comment (fpacifici) [reply tests/sentry/sentry_metrics/test_multiprocess_steps.py:297]: as long as this does not become part of the test suite, it is ok.

**Review by fpacifici — APPROVED** [review]

**Pull request merged**